src/util/pointer_offset_size.cpp

```diff
@@ -621,72 +599,102 @@ exprt build_sizeof_expr(
   return result;
 }
 
-bool get_subexpression_at_offset(
-  exprt &result,
-  mp_integer offset,
+exprt get_subexpression_at_offset(
+  const exprt &expr,
+  const mp_integer &offset,
   const typet &target_type_raw,
+  const irep_idt &byte_extract_op,
   const namespacet &ns)
 {
-  const typet &source_type=ns.follow(result.type());
-  const typet &target_type=ns.follow(target_type_raw);
+  if(offset == 0 && base_type_eq(expr.type(), target_type_raw, ns))
+  {
+    exprt result = expr;
+
+    if(expr.type() != target_type_raw)
+      result.type() = target_type_raw;
+
+    return result;
+  }
 
-  if(offset==0 && source_type==target_type)
-    return true;
+  const typet &source_type = ns.follow(expr.type());
+  const mp_integer target_size = pointer_offset_bits(target_type_raw, ns);
+  if(target_size < 0)
+    return nil_exprt();
 
   if(source_type.id()==ID_struct)
   {
-    const auto &st=to_struct_type(source_type);
-    const struct_typet::componentst &components=st.components();
-    member_offset_iterator offsets(st, ns);
-    while(offsets->first<components.size() &&
-          offsets->second!=-1 &&
-          offsets->second<=offset)
+    const struct_typet &struct_type = to_struct_type(source_type);
+
+    mp_integer m_offset_bits = 0;
+    for(const auto &component : struct_type.components())
     {
-      auto nextit=offsets;
-      ++nextit;
-      if((offsets->first+1)==components.size() || offset<nextit->second)
+      mp_integer m_size = pointer_offset_bits(component.type(), ns);
+      if(m_size < 0)
+        return nil_exprt();
+
+      if(
+        offset * 8 >= m_offset_bits && m_offset_bits % 8 == 0 &&
+        offset * 8 + target_size <= m_offset_bits + m_size)
```

Comment:
8?  Are we really hard-coded to have 8-bit bytes?

Reply:
Well, once #917 is merged (after me reworking some of it) that should no longer be the case.

Reply:
OK; I'll leave this one for now.